dj/construction/build.py

```diff
@@ -0,0 +1,275 @@
+                ast_tables = tables[table_node]
+                join_on = []
+                for table in ast_tables:
+                    for col in cols:
+                        join_on.append(
+                            ast.BinaryOp(
+                                ast.BinaryOpKind.Eq,
+                                ast.Column(ast.Name(col.name), _table=table),
+                                ast.Column(
+                                    ast.Name(col.dimension_column or "id"),
+                                    _table=dim_ast,
+                                ),
+                            ),
+                        )
+                select.from_.joins.append(  # pragma: no cover
+                    ast.Join(
+                        ast.JoinKind.LeftOuter,
+                        dim_ast,
+                        reduce(
+                            lambda left, right: ast.BinaryOp(
+                                ast.BinaryOpKind.And,
+                                left,
+                                right,
+                            ),
+                            join_on,
+                        ),
+                    ),
+                )
+
+    for node, tbls in tables.items():
+
+        if (
+            node.type != NodeType.SOURCE and build_plan_depth > 0
+        ):  # continue following build plan
+            _, node_build_plan = build_plan_lkp[node]
+            node_ast = node_build_plan[0]
+            node_query = build_query(
+                session,
+                node_ast,
+                node_build_plan,
+                build_plan_depth - 1,
+                database,
+                dialect,
+            )
+            alias = amenable_name(node.name)
+            node_select = node_query.select
+            node_ast = ast.Alias(ast.Name(alias), child=node_select)
+            for tbl in tbls:
+                select.replace(tbl, node_ast)
+        else:
+            node_table = [
+                table for table in node.tables if table.database.id == database.id
+            ][0]
+            node_ast = ast.Table(
+                ast.Name(node_table.table),
+                namespace=ast.Namespace(
+                    [
+                        ast.Name(s)
+                        for s in (node_table.catalog, node_table.schema_)
+                        if s
+                    ],
+                ),
+            )
+
+        for tbl in tbls:
+            select.replace(tbl, node_ast)
+
+    return select
+
+
+def build_query(  # pylint: disable=too-many-arguments
+    session: Session,
+    query: ast.Query,
+    build_plan: BuildPlan,
+    build_plan_depth: int,
+    database: Database,
+    dialect: Optional[str] = None,
+) -> ast.Query:
+    """transforms a query ast by replacing dj node references with their asts"""
+    select = query.to_select()
+    build_select(session, select, build_plan, build_plan_depth, database, dialect)
```

Comment:
It (like all the other steps in the pipeline) modifies the ast in-place. Do you think the return is deceptive?

Reply:
Yes, it fooled me. It's been ages since I modified input arguments in place through python functions purposefully. I'd like to hear what others think, but if you want to keep the structure of modifying in place then perhaps using a class and instance methods would be a compromise.

Reply:
That class suggestion from @agorajek gave me an idea. Could this be a method on the `Select` class? Maybe it can even just be a one-liner that calls this stand-alone function (although if it's only used in the context of an instance of Select, it should probably live as a method there). Then you can just call `query.to_select().build(session, build_plan, build_plan_depth, database, dialect)`

Reply:
In the end it would be nice to have something like:
```
query.to_select().build_for_database(database).serialize_for_engine(dialect)
```
or maybe even
```
query.build_for_database(database).serialize_for_engine(dialect)
```
?

Reply:
I think it's viable to add it as a method. I would still segregate this logic to a different file though don't you think and just call it in the method? I hope to do the same with the type inference e.g. allow calling `.type` on expression types

Reply:
Yeah I think a different file for the function makes sense and then just using that function in the method. That way if we start refactoring into smaller functions, as long as we keep that top level function signature we don't have to touch any classes.

Reply:
obfuscated the original functions in `build.py` with a leading `_`, removed return typing since it was confusing. added `.build` to `ast.Query`